### PR TITLE
Processors should be considered as rightful dependencies

### LINKED ISSUE #76: [BUG] Mapstruct processor dependency is deleted
**Describe the bug**
After depclean completes, and given mapstruct processor is used, the `mapstruct-processor` direct dependency is not longer present on the debloated `pom.xml`

**To Reproduce**
1. Configure the maven-processor-plugin to use the mapstruct annotation
```
      <plugin>
        <groupId>org.bsc.maven</groupId>
        <artifactId>maven-processor-plugin</artifactId>
        <executions>
          <execution>
            <id>process</id>
            <goals>
              <goal>process</goal>
            </goals>
            <phase>generate-sources</phase>
            <configuration>
              <processors>
                <processor>org.mapstruct.ap.MappingProcessor</processor>
              </processors>
            </configuration>
          </execution>
        </executions>
      </plugin>
```
2. Add required mapstruct dependencies
```
    <dependency>
      <groupId>org.mapstruct</groupId>
      <artifactId>mapstruct</artifactId>
    </dependency>
    <dependency>
      <groupId>org.mapstruct</groupId>
      <artifactId>mapstruct-processor</artifactId>
      <scope>provided</scope>
    </dependency>
```
3. Run depclean with debloated-pom generation

**Expected behavior**
`mapstruct-processor` dependency is conserved

**Additional context**
Mapstruct generates code during the `generated-sources` phase, but do not keep `mapstruct-processor` imports in the generated classes. This may be why it is eventually removed from the debloated-pom.

One could manually list this dependency into the `ignoreDependencies` configuration, but it may be better to consider annotation processor classes to be rightful `provided` dependencies

### PR DESCRIPTION
**Pull request by afillatre**
fixes #76

### CODE REVIEW

**Comment by afillatre** [comment]
Side-effect on the `artifactUsedClassesMap` attribut should be removed in another commit, but I modified a method outside my scope to be sure that at least the right state was returned. Please tell me if you're OK with that.

**Comment by cesarsotovalero** [comment]
> Side-effect on the `artifactUsedClassesMap` attribut should be removed in another commit, but I modified a method outside my scope to be sure that at least the right state was returned. Please tell me if you're OK with that.

Hi @afillatre, I'm OK with the proposal. I agree, the `DefaultProjectDependencyAnalyzer` needs some refactoring.

File: depclean-core/src/main/java/se/kth/depclean/core/analysis/DefaultProjectDependencyAnalyzer.java [hunk 1]
```diff
@@ -176,28 +224,21 @@ private void buildDependenciesDependencyClasses(MavenProject project) throws IOE
...
-          artifactUsedClassesMap.put(artifact, new HashSet<>());
-        }
-        artifactUsedClassesMap.get(artifact).add(clazz);
-        usedArtifacts.add(artifact);
-      }
+      findArtifactForClassName(artifactClassMap, clazz)
+          .ifPresent(artifact -> artifactUsedClassesMap.putIfAbsent(artifact, new HashSet<>()));
     }
-    return usedArtifacts;
+    return artifactUsedClassesMap.keySet();
```
Comment (cesarsotovalero) [inline depclean-core/src/main/java/se/kth/depclean/core/analysis/DefaultProjectDependencyAnalyzer.java:232]: Thanks for this refactoring! 
There are too many `null` in the project :)

File: depclean-maven-plugin/src/test/java/se/kth/depclean/DepCleanMojoIT.java [hunk 2]
```diff
@@ -19,11 +18,31 @@
...
+            "	com.fasterxml.jackson.core:jackson-core:2.12.2:compile (356 KB)",
+            "POTENTIALLY UNUSED DIRECT DEPENDENCIES [1]: ",
+            "	com.fasterxml.jackson.core:jackson-databind:2.12.2:compile (1 MB)",
+            "POTENTIALLY UNUSED INHERITED DEPENDENCIES [0]: ",
+            "POTENTIALLY UNUSED TRANSITIVE DEPENDENCIES [1]: ",
+            "	com.fasterxml.jackson.core:jackson-annotations:2.12.2:compile (73 KB)"
+    );
+  }
+
 }
```
Comment (cesarsotovalero) [inline depclean-maven-plugin/src/test/java/se/kth/depclean/DepCleanMojoIT.java:47]: Nice that you included a proper IT!

**Review by cesarsotovalero — APPROVED** [review]

**Review by cesarsotovalero — APPROVED** [review]

File: depclean-maven-plugin/src/test/java/se/kth/depclean/DepCleanMojoIT.java (hunk 2 above)
Comment (afillatre) [reply depclean-maven-plugin/src/test/java/se/kth/depclean/DepCleanMojoIT.java:47]: I had a very hard time debugging the plugin though, maybe I'm doing something wrong... I was unable to catch a breakpoint in the core project.

With more ITs, this big assertion can be refined, though.

**Pull request merged**